File: codes/86.分隔链表.cpp
/*
 * @lc app=leetcode.cn id=86 lang=cpp
 *
 * [86] 分隔链表
 */

// @lc code=start

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

/**
 * 思路：
 * 使用两个链表，一个放小于x的节点，一个放大于等于x的节点，将两个链表进行合并
*/

class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* small = new ListNode(-1);//构建虚拟头结点
        ListNode* p1 = small;//p1指针指向small链表末尾
        ListNode* large = new ListNode(-1);//构建虚拟头结点
        ListNode* p2 = large;//p2指针指向large链表末尾
        ListNode* p = head;//p指针遍历原链表
        while (p != nullptr){
            if (p->val >= x){//大于等于x的节点，放在large链表汇总
                p2->next = p;
                p2 = p2->next;
            }
            else{
                p1->next = p;
                p1 = p1->next;
            }
            ListNode* temp = p->next;//断开原链表中当前结点的next指针
            p->next = nullptr;
            p = temp;
        }
        //连接两个链表——small链表连接在large链表后面
        p1->next = large->next; 
        return small->next;
    }
};
// @lc code=end

File: codes/86.分隔链表.java
/*
 * @lc app=leetcode.cn id=86 lang=java
 *
 * [86] 分隔链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        // 核心原理是将一个链表分成两个链表
        // 两个链表的头节点
        ListNode head1 = new ListNode(-1);
        ListNode head2 = new ListNode(-1);
        // p1是小于x的数组成的链表，p2是大于等于x的数组成的链表
        ListNode p1 = head1, p2 = head2;
        ListNode p = head;
        while( p != null){
            if (p.val >= x){
                // 延长链表2
                p2.next = p;
                p2 = p2.next;
            }else{
                // 延长链表1
                p1.next = p;
                p1 = p1.next;
            }
            p = p.next;
            
        }
        // 将大链表接到小链表的后面
        // 前提是大链表的末尾要先指向空
        p2.next = null;
        p1.next = head2.next;
        return head1.next;
    }

}
// @lc code=end

